x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java

```diff
@@ -0,0 +1,391 @@
+                name(),
+                CoreValuesSourceType.KEYWORD,
+                cache,
+                breakerService,
+                (dv, n) -> new KeywordDocValuesField(FieldData.toString(dv), n)
+            ) {
+
+                @Override
+                public LeafOrdinalsFieldData load(LeafReaderContext context) {
+                    final SortedSetDocValues dvValues;
+                    final BinaryDocValues dvCounts;
+                    try {
+                        dvValues = DocValues.getSortedSet(context.reader(), getFieldName());
+                        dvCounts = DocValues.getBinary(context.reader(), countFieldType.name());
+                    } catch (IOException e) {
+                        throw new UncheckedIOException("Unable to load " + CONTENT_TYPE + " doc values", e);
+                    }
+
+                    return new AbstractLeafOrdinalsFieldData(toScriptFieldFactory) {
+
+                        @Override
+                        public SortedSetDocValues getOrdinalsValues() {
+                            return new CountedKeywordSortedBinaryDocValues(dvValues, dvCounts);
+                        }
+
+                        @Override
+                        public long ramBytesUsed() {
+                            return 0; // Unknown
+                        }
+
+                        @Override
+                        public void close() {
+                            // nothing to close
+                        }
+                    };
+                }
+
+                @Override
+                public LeafOrdinalsFieldData loadDirect(LeafReaderContext context) {
+                    return loadDirect(context);
+                }
+
+                @Override
+                public SortField sortField(
+                    Object missingValue,
+                    MultiValueMode sortMode,
+                    XFieldComparatorSource.Nested nested,
+                    boolean reverse
+                ) {
+                    throw new UnsupportedOperationException("can't sort on the [" + CONTENT_TYPE + "] field");
+                }
+
+                @Override
+                public BucketedSort newBucketedSort(
+                    BigArrays bigArrays,
+                    Object missingValue,
+                    MultiValueMode sortMode,
+                    XFieldComparatorSource.Nested nested,
+                    SortOrder sortOrder,
+                    DocValueFormat format,
+                    int bucketSize,
+                    BucketedSort.ExtraData extra
+                ) {
+                    throw new IllegalArgumentException("can't sort on the [" + CONTENT_TYPE + "] field");
+                }
+            };
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+    }
+
+    private static class CountedKeywordSortedBinaryDocValues extends AbstractSortedSetDocValues {
```

Comment:
The core of this field mapper is in this inner class. I think we should test this better.
In order to do that I think we should turn the `AbstractIndexOrdinalsFieldData` anonymous class into an inner class and write a test like `KeyedFlattenedLeafFieldDataTests`. In this test we should focus on what `CountedKeywordSortedBinaryDocValues ` is doing (iterating over duplicate values).

Reply:
Thanks for the hint. I started doing that but had the impression that this class would be rather hard to test (requires to mock Lucene classes) and mainly forwards its arguments to `CountedKeywordSortedBinaryDocValues`. So instead I tested that class in `CountedKeywordFieldTypeTests` (pushed in https://github.com/elastic/elasticsearch/pull/101826/commits/697adb42d83ea9cbb22314a265c57b1cf6f68745)